File: campaign-service/src/main/java/com/genphish/campaign/entity/TrackingEvent.java
package com.genphish.campaign.entity;

import com.genphish.campaign.entity.enums.TrackingEventType;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "tracking_events",
        indexes = {
                @Index(name = "idx_tracking_campaign", columnList = "campaign_id"),
                @Index(name = "idx_tracking_employee", columnList = "employee_id"),
                @Index(name = "idx_tracking_company", columnList = "company_id")
        })
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TrackingEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "campaign_id", nullable = false)
    private UUID campaignId;

    @Column(name = "employee_id", nullable = false)
    private UUID employeeId;

    @Column(name = "company_id", nullable = false)
    private UUID companyId; // Denormalized for fast multi-tenant queries

    @Enumerated(EnumType.STRING)
    @Column(name = "event_type", nullable = false)
    private TrackingEventType eventType; // EMAIL_OPENED, LINK_CLICKED, CREDENTIALS_SUBMITTED

    @Column(name = "occurred_at", nullable = false)
    private LocalDateTime occurredAt;

    @PrePersist
    protected void onCreate() {
        if (this.occurredAt == null) {
            this.occurredAt = LocalDateTime.now();
        }
    }
}

File: tracker-service/internal/models/tracking_event.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type TrackingEventType string

const (
	EventEmailOpened          TrackingEventType = "EMAIL_OPENED"
	EventLinkClicked          TrackingEventType = "LINK_CLICKED"
	EventCredentialsSubmitted TrackingEventType = "CREDENTIALS_SUBMITTED"
	EventDownloadTriggered    TrackingEventType = "DOWNLOAD_TRIGGERED"
	EventConsentGranted       TrackingEventType = "CONSENT_GRANTED"
)

type TrackingEvent struct {
	CampaignID uuid.UUID         `json:"campaignId"`
	EmployeeID uuid.UUID         `json:"employeeId"`
	CompanyID  uuid.UUID         `json:"companyId"`
	EventType  TrackingEventType `json:"eventType"`
	Timestamp  time.Time         `json:"timestamp"`
	UserAgent  string            `json:"userAgent,omitempty"`
	IPAddress  string            `json:"ipAddress,omitempty"`
}
